[SPARK-29500][SQL][SS] Support partition column when writing to Kafka

redsk opened the pull request:
### What changes were proposed in this pull request?
https://issues.apache.org/jira/browse/SPARK-29500

`KafkaRowWriter` now supports setting the Kafka partition by reading a "partition" column in the input dataframe.

Code changes in commit nr. 1.
Test changes in commit nr. 2.
Doc changes in commit nr. 3.

@tcondie @dongjinleekr @srowen 

### Why are the changes needed?
While it is possible to configure a custom Kafka Partitioner with
`.option("kafka.partitioner.class", "my.custom.Partitioner")`, this is not enough for certain use cases. See the Jira issue.

### Does this PR introduce any user-facing change?
No, as this behaviour is optional.

### How was this patch tested?
Two new UT were added and one was updated.

redsk committed:
[SPARK-29500] Add support for Kafka partition in KafkaWriter

redsk committed:
[SPARK-29500] Add support for Kafka partition in KafkaWriter (tests)

redsk committed:
[SPARK-29500] Add support for Kafka partition in KafkaWriter (doc)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -404,20 +420,85 @@ abstract class KafkaSinkBatchSuiteBase extends KafkaSinkSuiteBase {
     )
   }
 
+  test("batch - partition column vs default Kafka partitioner") {
```
Given we add this test to check the functionality to specify partition, I think we don't need to change above test.

redsk replied:
I did it because a previous PR (headers) changed it as well. But I can revert it to the previous state if so desired.

HeartSaVioR replied:
Personally yes, but we can hear more voices on this as well.

gaborgsomogyi replied:
I've gone through the test but not sure if it covers all. Here are my thought:
* Not sure why 100 partitions necessary?
* Don't we need the following tests? 1. `kafka.partitioner.class` overrides default 2. column overrides `kafka.partitioner.class` and default
* I think this test can be formed more simple. Put fixed key into a topic, get the partition (just like it is now but see my next comment). In another topic put all the data into a partition which is not the formerly get partition. Finally read back and double check.
* `.collect().toList.head` maybe not enough because it would be good to make sure the data is in one partition.
* There are couple of copy-pastes here as well.

redsk replied:
- "Not sure why 100 partitions necessary?": I just wanted to avoid the test to succeed by chance. I admit it's overkill. I'll reduce it to 10.
- "Don't we need the following tests?". See my other comment.
- "`.collect().toList.head` maybe not enough because it would be good to make sure the data is in one partition." `.collect()` docstring says "Returns an array that contains all rows in this Dataset. Running collect requires moving all the data into the application's driver process, and doing so on a very large dataset can crash the driver process with OutOfMemoryError." Does `collect()` work on a per-partition base? If so maybe I could use `.coalesce(1)` before calling `collect()`.
- "I think this test can be formed more simple." Ok it makes sense. I'll simplify it.
- "There are a couple of copy-pastes" Do you refer to the ```df.write
      .format("kafka")
      .option("kafka.bootstrap.servers", testUtils.brokerAddress)
      .option("topic", topic)
      .mode("append")
      .save()``` blocks? In that case, they are also all over the file.

redsk replied:
> I think `test("batch - write to kafka")` should be the simplest test and shouldn't contain neither headers nor partitions (I know for headers it's a bit late).

I can revert it to the previous state (with headers). Is that ok?

gaborgsomogyi replied:
> I can revert it to the previous state (with headers). Is that ok?

That's fine since that code is already in.

gaborgsomogyi replied:
> .collect() docstring says "Returns an array that contains all rows in this Dataset. Running collect requires moving all the data into the application's driver process, and doing so on a very large dataset can crash the driver process with OutOfMemoryError." Does collect() work on a per-partition base? If so maybe I could use .coalesce(1) before calling collect().

What I mean something like this:
`createKafkaReader(keyTopic).select("partition").map(_.getInt(0)).collect().toSet.size() === 1`

gaborgsomogyi replied:
> In that case, they are also all over the file.

Yes, the code can be enhanced several places but that doesn't mean we have to do the same copy-paste.

redsk replied:
> What I mean something like this:
> `createKafkaReader(keyTopic).select("partition").map(_.getInt(0)).collect().toSet.size() === 1`

I'm not sure I understand. Do you want me to assert that? Is this necessary considering that I send just one record?

redsk replied:
> Yes, the code can be enhanced several places but that doesn't mean we have to do the same copy-paste.

There are different opinions on copy-paste in UTs and I just copied the pattern in use in the file. But I also lean towards *not* copy-pasting. I can write a small function for that. Shall I apply it to other tests in the same function or just my contribution?

gaborgsomogyi replied:
In order not to have feature creep I would suggest to keep this for you change set. Later we can file another PR for the rest...

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -404,20 +420,85 @@ abstract class KafkaSinkBatchSuiteBase extends KafkaSinkSuiteBase {
 
+  test("batch - partition column vs default Kafka partitioner") {
+    val fixedKey = "fixed_key"
+    val nrPartitions = 100
```
It wouldn't matter if we just set number of partitions to 10 - we have flaky test failures where we cannot get metadata from Kafka in timeout, so I'd rather not make pressure on Kafka side.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -404,20 +420,85 @@ abstract class KafkaSinkBatchSuiteBase extends KafkaSinkSuiteBase {
+    )
+  }
+
+  test("batch - non-existing partitions trigger standard Kafka exception") {
```
Personally I would avoid depending on the behavior of Kafka - I'd feel we don't need to have this test, but let's hear more voices.

redsk replied:
My main question was: "what happens if I specify partitions that don't exist?" And the answer is that Kafka throws an exception with msg: "Topic $topic not present in metadata after $producerTimeout ms" which is not particularly clear. 

I could have called `producer.partitionsFor(topic)` to check if the specified partition is in range (and throw a more meaningful exception if not) but it's an expensive call for each message. I could cache the result but the number of partitions could change in between calls and I didn't want to complicate the implementation. Hence the need to see what the standard Kafka producer does (which is "not much").

gaborgsomogyi replied:
I think it's important to know what happens when such case happens. We've tested it and it's good.
From Spark test perspective we must ensure that our code works as expected and I think we don't have to pull in Kafka internals (I mean messages and timeouts) unless it's super important. I think it's not the case here which makes this test unnecessary from my point of view.

redsk replied:
Ok, I'll remove it.

gaborgsomogyi reviewed (commented):
Looks like a good feature.
* Just to double check won't this change the schema of a stateful query?
* The continuous test coverage can be increased

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -293,6 +293,21 @@ class KafkaSinkStreamingSuite extends KafkaSinkSuiteBase with StreamTest {
     assert(ex.getMessage.toLowerCase(Locale.ROOT).contains(
       "key attribute type must be a string or binary"))
+
+    try {
```
Actually I just submitted a patch #26158 to handle refactor to old ones: depending on which one to merge first, we can apply it.

HeartSaVioR commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala:
```diff
@@ -92,16 +92,19 @@ private[kafka010] abstract class KafkaRowWriter(
       throw new NullPointerException(s"null topic present in the data. Use the " +
         s"${KafkaSourceProvider.TOPIC_OPTION_KEY} option for setting a default topic.")
     }
+    val partition: java.lang.Integer =
```
It's to put `null` into parameter : Scala `Int` doesn't allow null.

gaborgsomogyi replied:
Makes sense, thanks. This case `Integer` is enough, right?

redsk replied:
`java.lang.Integer` allows `null` so it's enough. If I understood you.

gaborgsomogyi replied:
What I mean: `val partition: Integer =...`

HeartSaVioR commented:
> Just to double check won't this change the schema of a stateful query?

The `partition` attribute is only added on writer schema in this patch; it won't affect the schema of a stateful query. The `partition` attribute already exists in reader schema.

redsk commented on docs/structured-streaming-kafka-integration.md:
````diff
@@ -622,6 +626,10 @@ a ```null``` valued key column will be automatically added (see Kafka semantics
 is used as the topic when writing the given row to Kafka, unless the "topic" configuration
 option is set i.e., the "topic" configuration option overrides the topic column.
+If a partition column is not specified then the partition is calculated by the Kafka producer
+(using ```org.apache.kafka.clients.producer.internals.DefaultPartitioner```).
````
Sure, will do.

redsk commented on docs/structured-streaming-kafka-integration.md:
````diff
@@ -622,6 +626,10 @@ a ```null``` valued key column will be automatically added (see Kafka semantics
 option is set i.e., the "topic" configuration option overrides the topic column.
+If a partition column is not specified then the partition is calculated by the Kafka producer
+(using ```org.apache.kafka.clients.producer.internals.DefaultPartitioner```).
+This can be overridden in Spark by setting the ```kafka.partitioner.class``` option.
````
Yes, exactly. But this is `KafkaProducer` standard behaviour: 
- it uses `ProducerRecord` partition field. if `null`, fall backs to:
- `kafka.partitioner.class` provided. If not set:
- use default partitioner.

I don't believe we need a test for this (otherwise we would be testing Kafka API) but maybe we should explicitly state it in the doc.

gaborgsomogyi replied:
What I would like to test here is that Spark parameterizes producer properly.
These testing points has nothing to do with producer internals:
* `kafka.partitioner.class` config reaches producer instances and takes effect
* Partition field is set under some circumstances

gaborgsomogyi replied:
> Any config that starts with kafka. will be passed down to the producer, actually.

There are exceptions, please see them either in the doc or in the code. If such thing is not checked by a test then it can be broken easily. Just a personal thought: if something is not covered in a test later it's super hard to find out it's a bug or a feature...

> Don't the other test I added prove this point already?

If we add the mentioned simplified test then you're right it will cover it.

redsk replied:
Do we still need test 1 (`kafka.partitioner.class` overrides default partitioner)?

redsk commented:
>     * The continuous test coverage can be increased

What would you suggest? Aren't batch, streaming, and continuous streaming using the same `KafkaRowWriter` class under the hood?

gaborgsomogyi commented:
> What would you suggest? Aren't batch, streaming, and continuous streaming using the same KafkaRowWriter class under the hood?

I've had a deeper look and I tend to agree to depend on micro-batch tests.

redsk commented:
> I've had a deeper look and I tend to agree to depend on micro-batch tests.

So no more tests need on this front, right?

gaborgsomogyi commented:
Yep.

redsk committed:
[SPARK-29500] Rewritten tests

redsk commented:
@gaborgsomogyi I've addressed requests for doc and code. 

I've simplified the test `batch - partition column sets partition in kafka writes` as you suggested
I've added the test `batch - partition column and partitioner priorities` which shows the priorities of the different partitioning strategies.

Actually, I'm not sure the former test is still relevant as IMO completely subsumed by the second.

uncleGen commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -515,3 +619,15 @@ class KafkaSinkBatchSuiteV2 extends KafkaSinkBatchSuiteBase {
+                          value: Any,
+                          valueBytes: Array[Byte],
+                          cluster: Cluster
+                        )
```
nit: indent more or less.

redsk committed:
[SPARK-29500] Fixed indentation

gaborgsomogyi reviewed (commented):
Only minor things found.

gaborgsomogyi commented:
As we've spoken it would be good to merge https://github.com/apache/spark/pull/26158 before this to simplify the area a bit. So I'm going to give my approval when the other one merged.

redsk committed:
[SPARK-29500] Tests cleanup

redsk commented:
@gaborgsomogyi All changes made.

> As we've spoken it would be good to merge #26158 before this to simplify the area a bit. So I'm going to give my approval when the other one merged.

Does #26158 require more work? Doesn't seem so by looking at its thread.

Do you think there is any chance of a Spark 2.4.5 release? In that case, I'd happily backport this PR to the 2.4 branch.

gaborgsomogyi commented:
@redsk https://github.com/apache/spark/pull/26158 needs to be reviewed and I'm doing that right now.

> Do you think there is any chance of a Spark 2.4.5 release?

This is a new feature and we're not backporting such things so I don't think it will appear on 2.4 line.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -515,3 +591,14 @@ class KafkaSinkBatchSuiteV2 extends KafkaSinkBatchSuiteBase {
+      value: Any,
+      valueBytes: Array[Byte],
+      cluster: Cluster)
+  : Int = 0
```
nit: append this to the previous line

redsk replied:
According to the style guide, https://github.com/databricks/scala-style-guide#spacing-and-indentation , the return type on a new line is also acceptable (and I like it better :)

HeartSaVioR replied:
you can take a look at other codes and see how the code looks like. most likely we don't prefer unnecessary line.

gaborgsomogyi replied:
+1 sticking to the available code style.

redsk replied:
Alright then. Changed.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala:
```diff
@@ -418,6 +435,93 @@ abstract class KafkaSinkBatchSuiteBase extends KafkaSinkSuiteBase {
+    assert(partitions.head != keyPartition)
+  }
+
+  test("batch - partition column and partitioner priorities") {
```
Looks like this test covers everything. Great.

HeartSaVioR reviewed (approved):
LGTM except a nit.

HeartSaVioR commented:
I can even rebase #26158 with the change here if this can be merged sooner than #26158.

gaborgsomogyi commented:
> I can even rebase #26158 with the change here if this can be merged sooner than #26158.

Makes sense, either will be merged indicates merge on the other side.

gaborgsomogyi reviewed (approved):
LGTM. Minus the nit.

redsk committed:
[SPARK-29500] Code style

redsk commented:
It seems that linter checks in CI have been failing in the last 12 hours. 
```
[error] (*:update) sbt.ResolveException: unresolved dependency: com.typesafe#sbt-mima-plugin;0.3.0: not found
```
Nobody seems to have mentioned that in the Spark dev mailing list, though.

HeartSaVioR commented:
The issue regarding mima plugin got resolved.

redsk commented:
Is there anything else I can do? Do you thing this can be merged now?

HeartSaVioR commented:
Sorry, but I'm not a committer. You may still be encouraged to ping someone who "commits" the file (not author list) to review this. Like cc. someone.

HeartSaVioR commented:
cc. @tdas @zsxwing @jose-torres as they are committers handled many of things in SS area.

srowen reviewed (commented):
Looks OK pending tests

SparkQA commented:
**[Test build #4904 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4904/testReport)** for PR 26153 at commit [`96a8e9f`](https://github.com/apache/spark/commit/96a8e9f24fe7446ae444621cdb5c16e8d7c86a54).

SparkQA commented:
**[Test build #4904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4904/testReport)** for PR 26153 at commit [`96a8e9f`](https://github.com/apache/spark/commit/96a8e9f24fe7446ae444621cdb5c16e8d7c86a54).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

redsk commented:
@srowen it seems that tests passed cleanly. Is this PR ready to be merged?

srowen commented:
Merged to master
